app/services/liquid_cached_assigns.rb

```diff
@@ -0,0 +1,68 @@
+class LiquidCachedAssigns < Hash
+  AVAILABLE_PROJECT_ASSIGNS = %w{ evidences issues nodes notes project tags }.freeze
+
+  attr_accessor :assigns, :project
+
+  def initialize(project:)
+    @project = project
+
+    @assigns = { 'project' => ProjectDrop.new(project) }
+    @assigns.merge!(assigns_pro)
+  end
+
+  def [](record_type)
+    assigns[record_type] ||= cached_drops(record_type)
+  end
+
+  # SEE: https://github.com/Shopify/liquid/blob/77bc56/lib/liquid/context.rb#L211
+  # Liquid is checking if the variable is present in the assigns hash by
+  # calling the `key?` method. Since we're lazily loading the keys, the variable
+  # may not yet be present in the assigns hash.
+  def key?(key)
+    AVAILABLE_PROJECT_ASSIGNS.include?(key.to_s) || assigns.key?(key)
+  end
+
+  def merge(hash)
+    LiquidCachedAssigns.new(project: project).merge!(@assigns.merge(hash))
```

Comment:
isn't this f'ed up?

`LiquidCachedAssigns.new(project: project)` gives you exactly what you have in the context (`self`). Then you're calling merge on self's @assigns before passing to merge! which does it again (!?)

```
@assigns.merge!(hash)
```

at which point, why do you need `merge!`

Reply:
> LiquidCachedAssigns.new(project: project) gives you exactly what you have in the context (self).

Not quite as the new instance `LiquidCachedAssigns.new(project: project)` will not have the `self`'s current version of `@assigns`. If we remove the `@assigns` in the `merge` method:
```
def merge(hash)
  LiquidCachedAssigns.new(project: project).merge!(hash)
end
```
The following happens:
```
irb(main):003:0> assigns1 = LiquidCachedAssigns.new(project: Project.find(1))
irb(main):004:0> assigns1.merge!({'issue' => IssueDrop})
irb(main):005:0> assigns2 = assigns1.merge({'evidence' => EvidenceDrop})
irb(main):006:0> assigns2.assigns
=> {"project"=>ProjectDrop, "evidence"=>EvidenceDrop}
```
Notice that the `{ 'issue' => IssueDrop }` is missing. This is because `LiquidCachedAssigns.new(project: project)` is instantiated without the original's `@assigns` hash (merged with the IssueDrop in the above example).

---

I do see that the multiple `merge` calls can be confusing here so some possible alternatives to the approach are:

- Adding `assigns` to the initializer params so the merge method will look like this instead:
```
def merge(hash)
  LiquidCachedAssigns.new(assigns: @assigns, project: project).merge!(hash)
end
```
- Further abstraction by adding a `dup` method. The methods will look something like:
```
def dup
  LiquidCachedAssigns.new(project: project).merge!(@assigns)
end

def merge(hash)
  dup.merge!(hash)
end
```

@etdsoft thoughts?